sdk/storage/azure-storage-common/src/file_io.cpp

```diff
@@ -30,15 +30,15 @@ namespace Azure { namespace Storage { namespace _internal {
 #if defined(AZ_PLATFORM_WINDOWS)
   FileReader::FileReader(const std::string& filename)
   {
-    int sizeNeeded
-        = MultiByteToWideChar(CP_UTF8, MB_ERR_INVALID_CHARS, filename.data(), -1, nullptr, 0);
+    int sizeNeeded = MultiByteToWideChar(
+        CP_UTF8, MB_ERR_INVALID_CHARS, filename.data(), int(filename.length()), nullptr, 0);
     if (sizeNeeded == 0)
     {
       throw std::runtime_error("invalid filename");
     }
-    std::wstring filenameW(sizeNeeded, L'\0');
+    std::wstring filenameW(sizeNeeded + 1, L'\0');
```

Comment:
The `+ 1` is unnecessary here. `std::wstring` (as for all specializations of `std::basic_string`) is always null-terminated, and the null terminator isn't included in the size of the string. The effect of the `+ 1` here is that there are two `L'\0'` characters at the end of this string, one that is included in `filename.length()` and one that is not. (Also on line 88.)

Reply:
Yeah, I agree. But when we pass the string to another function, the function reads the characters in the string one by one, until meets a `\0`. The function needs to actually **read** the `\0` to know it is a terminator.

Isn't this out-of-range access an undefined behavior?

Reply:
No, because basic_string already puts a null there. http://eel.is/c++draft/basic.string#general-3

Reply:
fixed